File: src/main/java/rr/language/inspections/NonExistingUnit.java
package rr.language.inspections;

import com.intellij.codeInspection.LocalInspectionToolSession;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;
import rr.language.RRUtil;
import rr.language.Util;
import rr.language.psi.*;

import java.util.Collection;
import java.util.stream.Collectors;

public class NonExistingUnit extends Inspector {
    @NotNull
    protected RRVisitor buildRRVisitor(@NotNull final ProblemsHolder holder, @NotNull LocalInspectionToolSession session) {
        return new RRVisitor() {
            @Override
            public void visitUnitRef(@NotNull RRUnitRef element) {
                Collection<String> all = RRUtil.findAllUnitsAsStrings(element.getProject());

                if (!all.contains(element.getText())) {
                    holder.registerProblem(element, "Non existing unit", ProblemHighlightType.ERROR);
                }
            }

            public void visitStrUnitRef(@NotNull RRStrUnitRef element) {
                Collection<String> all = RRUtil.findAllUnitsAsStrings(element.getProject());

                if (!all.contains(Util.unquote(element.getText()))) {
                    holder.registerProblem(element, "Non existing unit", ProblemHighlightType.ERROR);
                }
            }
        };
    }
}

File: src/main/java/rr/language/inspections/NonExistingUnit.kt
package rr.language.inspections

import com.intellij.codeInspection.LocalInspectionToolSession
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import rr.language.RRUtil.findAllUnitsAsStrings
import rr.language.Util.unquote
import rr.language.psi.RRStrUnitRef
import rr.language.psi.RRUnitRef
import rr.language.psi.RRVisitor

class NonExistingUnit : Inspector() {
    override fun buildRRVisitor(holder: ProblemsHolder, session: LocalInspectionToolSession): RRVisitor {
        return object : RRVisitor() {
            override fun visitUnitRef(element: RRUnitRef) {
                val all = findAllUnitsAsStrings(element.project)
                if (!all.contains(element.text)) {
                    holder.registerProblem(element, "Non existing unit", ProblemHighlightType.ERROR)
                }
            }

            override fun visitStrUnitRef(element: RRStrUnitRef) {
                val all = findAllUnitsAsStrings(element.project)
                if (!all.contains(unquote(element.text))) {
                    holder.registerProblem(element, "Non existing unit", ProblemHighlightType.ERROR)
                }
            }
        }
    }
}